src/lib.rs

```diff
@@ -0,0 +1,1210 @@
+        let mut v = Vob::new();
+        v.extend(iter);
+        v
+    }
+}
+
+
+// This is based on the `BitVec` approach to indices. It's clearly a horrible way of doing things,
+// but until `IndexGet` is implemented, we're stuck.
+static TRUE: bool = true;
+static FALSE: bool = false;
+
+impl<T: Debug + One + PrimInt + Zero> Index<usize> for Vob<T> {
+    type Output = bool;
+
+    fn index(&self, index: usize) -> &bool {
+        match self.get(index) {
+            Some(true) => &TRUE,
+            Some(false) => &FALSE,
+            None => panic!("index out of bounds: the len is {} but the index is {}", self.len, index)
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Iter<'a, T: 'a> {
+    vob: &'a Vob<T>,
+    range: Range<usize>,
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> Iterator for Iter<'a, T> {
+    type Item = bool;
+
+    fn next(&mut self) -> Option<bool> {
+        self.range.next()
+                  .map(|i| self.vob.get(i).unwrap())
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.range.size_hint()
+    }
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> DoubleEndedIterator for Iter<'a, T> {
+    fn next_back(&mut self) -> Option<bool> {
+        self.range.next_back()
+                  .map(|i| self.vob.get(i).unwrap())
+    }
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> ExactSizeIterator for Iter<'a, T> { }
+
+impl<'a, T: Debug + One + PrimInt + Zero> IntoIterator for &'a Vob<T> {
+    type Item = bool;
+    type IntoIter = Iter<'a, T>;
+
+    fn into_iter(self) -> Iter<'a, T> {
+        self.iter()
+    }
+}
+
+#[derive(Clone)]
+pub struct IterSetBits<'a, T: 'a> {
+    vob: &'a Vob<T>,
+    range: Range<usize>,
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> Iterator for IterSetBits<'a, T> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<usize> {
+        if let Some(mut i) = self.range.next() {
+            for b in block_offset::<T>(i) .. blocks_required::<T>(self.vob.len) {
+                let v = self.vob.vec[b];
+                if v != T::zero() {
+                    let i_off = i % bits_per_block::<T>();
+                    let tz = (v >> i_off).trailing_zeros() as usize;
+                    if tz < bits_per_block::<T>() {
+                        let bs = b * bits_per_block::<T>() + i_off + tz;
+                        self.range.start = bs + 1;
+                        return Some(bs);
+                    }
+                }
+                i = b * bits_per_block::<T>();
+            }
+        }
+        self.range.start = self.range.end;
+        None
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.range.size_hint()
+    }
+}
+
+#[derive(Clone)]
+pub struct IterUnsetBits<'a, T: 'a> {
+    vob: &'a Vob<T>,
+    range: Range<usize>,
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> Iterator for IterUnsetBits<'a, T> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<usize> {
```

Comment:
Would it be easier to implement this in terms of `IterSetBits`?

Reply:
I don't think so (at least, I can't work out how!).

Reply:
Well, if you had a `SetBitsIter` and it first gave you `6`, you'd know bits 1-5 are not set. then if it gave you `8`, you'd know `7` was not set.

It's just an idea. If it's not good, let's move on.

Reply:
Yes, but that's a declarative specification (actually, specification by example ;) ), not an algorithm. If you think about the work needed to turn it into an algorithm, it's clearly more work than the existing approach, because it's a superset of the work needed. Conclusion: I don't think it's a winner :)